### DISCUSSION TITLE
HA Topologies for Artemis.

### DISCUSSION THREAD

**Discussion by achinjain214** [arkmq-org/arkmq-org-broker-operator#681] [Q&A]
What type of HA topologies are available for Artemis MQ ? Also which topology is best for replication ?

So how the sync work between the network of brokers ?

**Comment by brusdev** [comment]
The HA on Kubernetes/OpenShift is provided by the platform itself and the ArtemisCloud Operator supports message migration if you scale down your cluster. For generic questions related to ActiveMQ Artemis HA you can send an email to users@activemq.apache.org

**Reply by achinjain214** [reply]
So if available zone itself goes down ( on which broker pod is running) with which EBS volume will not be available , so  how will message migration will work in this scenario?

**Reply by achinjain214** [reply]
@brusdev we are planning to use current Artemis operator model for Production , is it possible we can connect with you for a quick catchup whenever you have time . Any help would be much appreciated.

Thanks
Achin

**Reply by brusdev** [reply]
@achinjain214 if the persistent volume is not available then the message migration doesn't work. For a disaster recovery you could use ActiveMQ Artemis Mirroring, see https://activemq.apache.org/components/artemis/documentation/latest/amqp-broker-connections.html#mirroring

If you have any other question feel free to create new discussions. This is a great way to contribute to this open source project because also other users can benefit of the answers.

**Reply by andrewmcguinness** [reply]
ActiveMQ supports Live-Backup HA with replication - https://activemq.apache.org/components/artemis/documentation/latest/ha.html#high-availability-and-failover 

How do we ensure redundant storage of queued messages in Kubernetes?

**Reply by achinjain214** [reply]
@brusdev so Mirroring can be enabled on broker.xml on operator model ( EKS) using CRDS/CRs . I was trying to edit broker.xml but did not find any way via CRDs or CRs to implement replication or HA policy ( Master/Slave) in broker.xml.

Please let me know if any ways to edit broker.xml on EKS .

Thanks for all your help and support.

**Reply by brusdev** [reply]
> How do we ensure redundant storage of queued messages in Kubernetes?

It up to the Kubernetes admins ensure redundant storage (i.e. using [ceph](https://docs.ceph.com/)) for the persistent volume claims, this should be one of benefits to run an application on Kubernetes.

**Reply by brusdev** [reply]
> @brusdev so Mirroring can be enabled on broker.xml on operator model ( EKS) using CRDS/CRs . I was trying to edit broker.xml but did not find any way via CRDs or CRs to implement replication or HA policy ( Master/Slave) in broker.xml.

ActiveMQ Artemis Mirroring could be enabled using the field spec.brokerProperties, I haven't tried yet but there is a test in the ActiveMQ Artemis sources, see https://github.com/apache/activemq-artemis/blob/2.31.0/artemis-server/src/test/java/org/apache/activemq/artemis/core/config/impl/ConfigurationImplTest.java#L743

**Comment by FabrizioFubelli** [comment]
About HA Topologies, I'm looking for an `active-active` clustering configuration (all the instances should share the messages with each other).

I'm in OpenShift and I'm using a **Service** for load-balancing the traffic, but it happens that the **consumer** connects to one instance, while the **producer** writes on the other. It would be great if all the **ActiveMQ** replicas would be synchronized via the file system.

I can't configure the clients by using static host names, I can just use the **Service** name, which directs traffic towards replicas.

In this situation, what is the best configuration, to guarantee the `high-availability` ?

**Reply by gtully** [reply]
There is a proof of concept for pub_sub in one of the tests that I will link below.
The idea is that producers have the lowest latency HA requirements, they expect to be able to produce to the broker service. Consumers are happy to be event driven and can wait for their partition to restart.

Have N brokers, allow producers to connect to any broker. Addresses are mirrored (store and forward) such that a message produced on broker 1 is also produced on broker N.
Partition consumers across brokers (the connection router based on roles does this). Clients support retry/reconnect and eventually find a broker partition that will accept them, or keep trying if their partition is offline.

This scheme can scale nicely, the brokers are statically configured and not "clustered", they are really dumb/simple. There is very little bookkeeping or synchronisation overhead when an instance restarts.

we are working towards driving auto scaling for this sort of architecture from the operator.

Note that this test pulls together lots of configuration (jaas, rbac, mirror, router etc), there is lots to parse, but it captures a use case and simple architecture that can work today to achieve a simple HA broker service.
https://github.com/artemiscloud/activemq-artemis-operator/blob/main/controllers/activemqartemis_pub_sub_scale_test.go#L55  

For point-to-point - queue semantics, something similar with queue federation can work really well but this initial focus is on pub-sub. The key difference with queues is you typically want to collocate producers/consumers for addresses to respect data gravity so it is not as trivial to scale.

**Comment by Pandi-A** [comment]
@brusdev ,

We are planing to deploy Artemis CRD, we are trying to redistribute the messages between brokers, by default redistribution is false. How to enable the message redistribution

https://activemq.apache.org/components/artemis/documentation/latest/clusters.html#message-redistribution

So sender connected to broker 1 and consumer (Camel-K) connected to broker2, in this case consumer will not see any message. For avoiding this issue we are planning to redistribute it.

**Reply by brusdev** [reply]
@Pandi-A could you move this question to a new discussion?
